docs: how to add new (frontend) components (#847)
widgetti/solara: solara/website/pages/documentation/advanced/content/10-howto/70-make-a-vue-component.md

# Making a Vue based component

If based on [./make-components](./make-components) you made the decision to write a vue based component, this article will guide you through the process.

This howto is an extended version of [https://solara.dev/documentation/examples/general/vue_component]

> [https://solara.dev/documentation/api/utilities/component_vue]
> 

Reviewer: ```suggestion
```

> We will first start with a component which is meant to be use inside of a single application, which will fetch the 3rd party library from a CDN. The goal

Reviewer: ```suggestion
We will start by making a component that shows a button, which shows confetti when clicked. The component is meant to be use inside of a single application, and will fetch a 3rd party library from a CDN. The goal
```

is to get something work with the minimal amount of effort.

> Our goal is to create a button, that when clicked, will show confetti.

Reviewer: ```suggestion
```

## The skeleton setup.

First, we create our minimal `button-confetti.vue` file which includes a very simple button with a custom label and a click event handler.
```vue
<template>
    <button class="button-confetti" @click="click({extra: 'foo'}) ">
        \{\{ label \}\}
    </button>
</template>
```

To expose this vue component to Solara, we use [the component_vue decorator](/documentation/api/utilities/component_vue).

```python
import solara
from typing import Callable

@solara.component_vue("button-confetti.vue")
def ButtonConfetti(
    label: str = "Default label",
    event_click: Callable = None,
    ):
    ...

@solara.component
def Page():
   ButtonConfetti(label="Click me", event_click=print)
```

This simple app will show a button with the label "Click me" and when clicked, it will print `{'extra': 'foo'}` to the console. [<img src="https://py.cafe/logos/pycafe_logo.png" alt="PyCafe logo" width="24" height="24"> Run and edit this example at PyCafe](https://py.cafe/maartenbreddels/solara-howto-component-vue-A).

## Loading the confetti library.

> Now that we have our skeleton setup, we can add the confetti library. We will use the [`canvas-confetti`](https://www.npmjs.com/package/canvas-confetti) library, which is available on a CDN. We will add this to our `button-confetti.vue` file with unfortunately a bit of boilerplate code to load the library.:

Reviewer: ```suggestion
Now that we have our skeleton setup, we can add the confetti library. We will use the [`canvas-confetti`](https://www.npmjs.com/package/canvas-confetti) library, which is available on a CDN. We will add this to our `button-confetti.vue` file using a bit of boilerplate code:
```

```vue
<template>
    <button class="button-confetti" @click="clickHandler">
        {{label}}
    </button>
</template>
<script>
module.exports = {
    created() {
        console.log("button-confetti: created");
        this.loadedConfetti = this.loadConfetti();
    },
    mounted() {
        console.log("button-confetti: mounted");
    },
    watch: {
        label() {
            console.log("button-confetti: label changed");
        }
    },
    methods: {
        clickHandler() {
            if(this.click) {
                this.click({extra: 'foo'})
            }
            this.showConfetti();
        },
        async showConfetti() {
            // make sure it is loaded by waiting on the Promise
            await this.loadedConfetti;

>             await new Promise((resolve) => setTimeout(resolve, 1000))

Reviewer: I'm guessing this should either be removed or somehow commented on?

```suggestion
```

            require.config({
                map: {
                    '*': {
                        'confetti': confettiUrl,
                    }
                }
            })
            await new Promise((resolve, reject) => {
                require(['confetti'], (confettiModule) => {
                    resolve(confettiModule);
                }, reject)
            });
        },
        /* end boilerplate */
    },
}
</script>
<style id="button-confetti">
    .button-confetti {
        background-color: lightblue;
        margin: 10px;
        padding: 4px;
        border: 1px solid black;
    }
</style>
```

[<img src="https://py.cafe/logos/pycafe_logo.png" alt="PyCafe logo" width="24" height="24"> Run and edit this example at PyCafe](https://py.cafe/maartenbreddels/solara-howto-component-vue-B).

## Triggering the confetti from the Python side

Although we have a working button showing out confetti, this is triggered by a button in our vue template. There are situations where we want to trigger the confetti from the Python side. We can do this by making our

> Vue component respond to its argument in the `watch` section. We also remove any vue from out template so that our vue template becomes a non-visual component.

Reviewer 2: ```suggestion
Vue component respond to its argument in the `watch` section. We also remove any vue from our template so that our vue template becomes a non-visual component.
```
